File: src/swift/3_StructuralPatterns/3_Composite/GeometricShapes.swift
import Foundation

class GraphicObject : CustomStringConvertible
{
  var name: String = "Group"
  var color: String = ""

  var children = [GraphicObject]()

  init()
  {

  }

  init(name: String)
  {
    self.name = name
  }

  private func print(buffer: inout String, depth: Int)
  {
    buffer.append(String(repeating: "*", count: depth))
    buffer.append(color.isEmpty ? "" : "\(color) ")
    buffer.append("\(name)\n")

    for child in children
    {
      child.print(buffer: &buffer, depth: depth+1)
    }
  }

  var description: String
  {
    var buffer = ""
    print(buffer: &buffer, depth: 0)
    return buffer
  }
}

class Circle : GraphicObject
{
  init(color: String)
  {
    super.init(name: "Circle")
    self.color = color
  }
}

class Square : GraphicObject
{
  init(color: String)
  {
    super.init(name: "Square")
    self.color = color
  }
}

func main()
{
  let drawing = GraphicObject(name: "My Drawing")
  drawing.children.append(Square(color: "Red"))
  drawing.children.append(Circle(color: "Yellow"))

  let group = GraphicObject()
  group.children.append(Circle(color: "Blue"))
  group.children.append(Square(color: "Blue"))

  drawing.children.append(group)

  print(drawing.description)
}

main()

File: src/js/3_StructuralPatterns/3_Composite/geometric_shapes.js
class GraphicObject
{
  get name()
  {
    return this._name;
  }

  constructor(name='Group ' + (GraphicObject.count++))
  {
    this.children = [];
    this.color = undefined;
    this._name = name;
  }

  print(buffer, depth)
  {
    buffer.push('*'.repeat(depth));
    if (depth > 0)
      buffer.push(' ');
    if (this.color)
      buffer.push(this.color + ' ');
    buffer.push(this.name);
    buffer.push('\n');

    for (let child of this.children)
      child.print(buffer, depth+1);
  }

  toString()
  {
    let buffer = [];
    this.print(buffer, 0);
    return buffer.join('');
  }
}
GraphicObject.count = 0;

class Circle extends GraphicObject
{
  constructor(color)
  {
    super('Circle');
    this.color = color;
  }
}

class Square extends GraphicObject
{
  constructor(color)
  {
    super('Square');
    this.color = color;
  }
}

let drawing = new GraphicObject();
drawing.children.push(new Square('Red'));
drawing.children.push(new Circle('Yellow'));

let group = new GraphicObject();
group.children.push(new Circle('Blue'));
group.children.push(new Square('Blue'));
drawing.children.push(group);

console.log(drawing.toString());